native/python/src/fairseq2n/bindings/data/iterator_data_source.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/bindings/data/iterator_data_source.h"
+
+namespace py = pybind11;
+
+namespace fairseq2n::detail {
+
+std::optional<data>
+iterator_data_source::next()
+{
+    py::gil_scoped_acquire acquire;
+
+    if (reset_) {
+
+        reloaded_ = false;
+        reset_ = false;
+        *iterator_ = reset_fn_(*iterator_);
+        ++*iterator_;
+
+    } else if (reloaded_) {
+        // Saving/reloading the iterator may skip over an example,
+        // so we check if this iterator has been reloaded and 
+        // return the potentially missing example here.
+
+        if (to_return_) {
+            reloaded_ = false;
+        }
+        return to_return_;
+
+    }
+
+    if (*iterator_ == py::iterator::sentinel()) {
+        return std::nullopt;
+    }
+    return (*iterator_)++->cast<py_object>();
+}
+
+void
+iterator_data_source::reset(bool) noexcept
+{
+    reset_ = true;
+}
+
+void
+iterator_data_source::record_position(tape &t, bool) const
+{
+    py::gil_scoped_acquire acquire;
+
+    std::optional<data> to_return;
+    if (*iterator_ != py::iterator::sentinel()) {
+        to_return = (*iterator_)->cast<py_object>();
+    }
+
+    py::function pickle_dump_fn = py::module::import("pickle").attr("dumps");
+    t.record(pickle_dump_fn(*iterator_).cast<py_object>());
+
+    t.record(to_return);
+
+    t.record(reset_);
+}
+
+void
+iterator_data_source::reload_position(tape &t, bool)
```

Comment:
I think we need to add into the doc string explicitly that iterator should be picklable (and correctly) since the pipeline state will be serialized using pickle (which can make weird thing sometimes). 
I wonder if we do want to add some custom function for state_load/state_save along with `reset_fn`.
One different approach that could make sense in some case (where the elements production is fast but serialization can be complicated) to restore the state is to do `reset` + `skip` the number of element that were already produced.

Reply:
>I think we need to add into the doc string explicitly that iterator should be picklable (and correctly) since the pipeline state will be serialized using pickle (which can make weird thing sometimes).

PR forthcoming for this---the plan is to add a allow_unpickleable flag that is False by default (in which case it will raise an error on `__init__`), if that sounds good.

>I wonder if we do want to add some custom function for state_load/state_save along with reset_fn.
One different approach that could make sense in some case (where the elements production is fast but serialization can be complicated) to restore the state is to do reset + skip the number of element that were already produced.

I believe the user can already do this through `__getstate__` and `__setstate__` methods in the iterator class. For example:

```python
from fairseq2.data import read_iterator
from typing import Iterator

class A(Iterator):
    def __init__(self):
        self.counter = 0
        self.generator = iter(range(100))
    def __iter__(self):
        return self
    def __next__(self):
        self.counter += 1
        return next(self.generator)
    def __getstate__(self):
        return self.counter
    def __setstate__(self, counter):
        self.generator = iter(range(100))
        for i in range(counter):
            next(self.generator)
        self.counter = counter

pipeline = read_iterator(A(), reset_fn = lambda x : A(), infinite=False).and_return()
it = iter(pipeline)
print(next(it), next(it), next(it))
d = pipeline.state_dict()
print(next(it), next(it), next(it))
pipeline.load_state_dict(d)
print(next(it), next(it), next(it))
``` 
unless you'd prefer this built-in.